### DISCUSSION TITLE
Adding displayed text expansion

### DISCUSSION THREAD

**Discussion by bonnefoi** [kovidgoyal/kitty#5237] [Ideas]
Hello,

I'm using Kitty as my everyday terminal. 
Recently, I've found a much needed capability in xterm : `dabbrev-expand()` that could expand the word before the cursor by searching in the preceding text on the screen.

Do you think that such a feature could be added to kitty ?

best regards,
pfb

**Comment by page-down** [comment]
You can use a custom kitten to implement similar functionality. Here is an example of using hints kitten as the text selection UI.

dynamic_abbrev_hints.py

```python
import string
from typing import Any, Dict, List

from kitty.boss import Boss
from kittens.tui.handler import result_handler


def main(args: List[str]) -> None:
    pass


@result_handler(no_ui=True)
def handle_result(args: List[str], data: str, target_window_id: int, boss: Boss) -> None:
    w = boss.active_window
    if w is None:
        return
    s = w.screen
    if s.cursor.x == 0:
        return
    line = str(s.visual_line(s.cursor.y))
    chars = string.ascii_uppercase + string.ascii_lowercase + string.digits + '-_.'
    x = s.cursor.x
    while x > 0 and line[x - 1] in chars:
        x -= 1
    if x == s.cursor.x:
        return
    word = line[x:s.cursor.x]

    def done(data: Dict[str, Any], target_window_id: int, self: Boss) -> None:
        if data and data.get('match'):
            m = data['match'][0]
            w = boss.window_id_map.get(target_window_id)
            if w is not None:
                w.write_to_child(m[len(word):])

    boss.run_kitten_with_metadata('hints', args=('--type', 'regex', '--regex', word.replace('.', '\.') + '[a-zA-Z0-9-_.]*'), custom_callback=done)
```

```shell
kitty -o 'map f1 kitten dynamic_abbrev_hints.py'
```

Type some text and press F1. After selecting the matching text, the remaining characters will be filled in.
If kitty has a floating window feature in the future, it will be possible to list candidates directly.

https://sw.kovidgoyal.net/kitty/kittens/custom/

The above is just one way to do it, but you can actually just use the `hints --customize-processing` argument, provide a python script, and customize the `mark` and `handle_result` functions.
https://sw.kovidgoyal.net/kitty/kittens/hints/#completely-customizing-the-matching-and-actions-of-the-kitten

**Reply by bonnefoi** [reply]
Thank you very much for your code ! 
I've to install a newer version of kitty to make it work on my Ubuntu 22.04. I've also added the `:` in your regex to accomodate for IPv6 addresses.
I'm impressed by the new improvements of kitten based on a Python internal interpreter !

**Comment by kovidgoyal** [comment]
Feel free to do so, it should be a pretty simple script using kitty
remote control features or by writing a kitten
